service/src/main/java/crawlercommons/urlfrontier/service/AbstractFrontierService.java

```diff
@@ -678,11 +696,19 @@ public void getURLs(GetParams request, StreamObserver<URLInfo> responseObserver)
                 numQueuesTried,
                 requestID.toString());
 
-        getURLs_urls_count.inc(totalSent);
+        getURLs_urls_count.inc(totalSent.get());
 
         requestTimer.observeDuration();
 
-        responseObserver.onCompleted();
+        // wait for all threads to have finished
+        while (inProcess.get() != 0) {
```

Comment:
Wonder, if we can avoid active waiting here by using a `CountDownLatch` ?

Reply:
I don't think it is possible to increment the value of a CountDownLatch

Reply:
I was only looking at:

```java
  inProcess.incrementAndGet();
   // do something
  inProcess.decrementAndGet();
```
but we cannot pre-computed the init value of a latch, right?

Reply:
Forget that comments. I had some weird mind fog ;-)
You are right!

Reply:
> Forget that comments. I had some weird mind fog ;-) You are right!

Welcome to my world ;-)